In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv(r"E:\Data Base\open_pubs.csv",header = None)
df

0                                         1  \
0          22                                Anchor Inn   
1          36                        Ark Bar Restaurant   
2          74                                 Black Boy   
3          75                               Black Horse   
4          76                                Black Lion   
...       ...                                       ...   
51326  597119  Wrexham & District War Memorial Club Ltd   
51327  597127                 Wrexham Lager Social Club   
51328  597130         Wrexham Rail Sports & Social Club   
51329  597131                        Wrexham Rugby Club   
51330  597137                                     Y Tai   

                                                       2         3       4  \
0            Upper Street, Stratford St Mary, COLCHESTER   CO7 6LW  604749   
1      Ark Bar And Restaurant, Cattawade Street, Bran...  CO11 1RH  610194   
2      The Lady Elizabeth, 7 Market Hill, SUDBURY, Su...  CO10 2EA  587334   
3            Lower Street, Stratford St Mary, COLCHESTER   CO7 6JS  622675   
4                          Lion Road, Glemsford, SUDBURY  CO10 7RF  622675   
...                                                  ...       ...     ...   
51326  Wrexham War Memorial Club Farndon Street, Wrex...  LL13 8DE  333909   
51327                 1 - 3 Union Road, Wrexham, Wrexham  LL13 7SR  333028   
51328  Sports And Social Club, 44 Brook Street, Wrexh...  LL13 7LU  333259   
51329  Wrexham Rugby Club Bryn Estyn Road, Wrexham, W...  LL13 9TY  335808   
51330       Y Tai Railway Road, Brymbo, Wrexham, Wrexham  LL11 5EA  329721   

             5          6          7        8  
0       234404  51.970379   0.979340  Babergh  
1       233329  51.958698   1.057832  Babergh  
2       241316  52.038595   0.729915  Babergh  
3     -5527598         \N         \N  Babergh  
4     -5527598         \N         \N  Babergh  
...        ...        ...        ...      ...  
51326   350438  53.047100  -2.987319  Wrexham  
51327   350563  53.048123  -3.000485  Wrexham  
51328   350213  53.044998  -2.996966  Wrexham  
51329   351078  53.053094  -2.959124  Wrexham  
51330   353785  53.076638  -3.050512  Wrexham  

[51331 rows x 9 columns]

In [3]:
df.columns=['Fsa_id','Name','Address','Postcode','Easting','Northing','Latitude','Longitude','Local_authority']
df.head()

Fsa_id                Name  \
0      22          Anchor Inn   
1      36  Ark Bar Restaurant   
2      74           Black Boy   
3      75         Black Horse   
4      76          Black Lion   

                                             Address  Postcode  Easting  \
0        Upper Street, Stratford St Mary, COLCHESTER   CO7 6LW   604749   
1  Ark Bar And Restaurant, Cattawade Street, Bran...  CO11 1RH   610194   
2  The Lady Elizabeth, 7 Market Hill, SUDBURY, Su...  CO10 2EA   587334   
3        Lower Street, Stratford St Mary, COLCHESTER   CO7 6JS   622675   
4                      Lion Road, Glemsford, SUDBURY  CO10 7RF   622675   

   Northing   Latitude Longitude Local_authority  
0    234404  51.970379  0.979340         Babergh  
1    233329  51.958698  1.057832         Babergh  
2    241316  52.038595  0.729915         Babergh  
3  -5527598         \N        \N         Babergh  
4  -5527598         \N        \N         Babergh

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51331 entries, 0 to 51330
Data columns (total 9 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Fsa_id           51331 non-null  int64 
 1   Name             51331 non-null  object
 2   Address          51331 non-null  object
 3   Postcode         51331 non-null  object
 4   Easting          51331 non-null  int64 
 5   Northing         51331 non-null  int64 
 6   Latitude         51331 non-null  object
 7   Longitude        51331 non-null  object
 8   Local_authority  51331 non-null  object
dtypes: int64(3), object(6)
memory usage: 3.5+ MB


- from the above info, we can say that there is no null values.
- and the data frame has 51331 rows and 9 columns.

In [6]:
df['Latitude'] = df['Latitude'].replace(['\\N'],np.NaN)
df['Longitude'] = df['Longitude'].replace(['\\N'],np.NaN)
df.head()

Fsa_id                Name  \
0      22          Anchor Inn   
1      36  Ark Bar Restaurant   
2      74           Black Boy   
3      75         Black Horse   
4      76          Black Lion   

                                             Address  Postcode  Easting  \
0        Upper Street, Stratford St Mary, COLCHESTER   CO7 6LW   604749   
1  Ark Bar And Restaurant, Cattawade Street, Bran...  CO11 1RH   610194   
2  The Lady Elizabeth, 7 Market Hill, SUDBURY, Su...  CO10 2EA   587334   
3        Lower Street, Stratford St Mary, COLCHESTER   CO7 6JS   622675   
4                      Lion Road, Glemsford, SUDBURY  CO10 7RF   622675   

   Northing   Latitude Longitude Local_authority  
0    234404  51.970379  0.979340         Babergh  
1    233329  51.958698  1.057832         Babergh  
2    241316  52.038595  0.729915         Babergh  
3  -5527598        NaN       NaN         Babergh  
4  -5527598        NaN       NaN         Babergh

In [9]:
df.isnull().sum()

Fsa_id               0
Name                 0
Address              0
Postcode             0
Easting              0
Northing             0
Latitude           767
Longitude          767
Local_authority      0
dtype: int64

In [10]:
df = df.dropna()
df.head()

Fsa_id                Name  \
0      22          Anchor Inn   
1      36  Ark Bar Restaurant   
2      74           Black Boy   
5      97        Brewers Arms   
6     102        Bristol Arms   

                                             Address  Postcode  Easting  \
0        Upper Street, Stratford St Mary, COLCHESTER   CO7 6LW   604749   
1  Ark Bar And Restaurant, Cattawade Street, Bran...  CO11 1RH   610194   
2  The Lady Elizabeth, 7 Market Hill, SUDBURY, Su...  CO10 2EA   587334   
5  The Brewers Arms, Bower House Tye, Polstead, C...   CO6 5BZ   598743   
6                     Bristol Hill, Shotley, IPSWICH   IP9 1PU   624624   

   Northing   Latitude Longitude Local_authority  
0    234404  51.970379  0.979340         Babergh  
1    233329  51.958698  1.057832         Babergh  
2    241316  52.038595  0.729915         Babergh  
5    240655  52.028694  0.895650         Babergh  
6    233550  51.955042  1.267642         Babergh

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 50564 entries, 0 to 51330
Data columns (total 9 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Fsa_id           50564 non-null  int64 
 1   Name             50564 non-null  object
 2   Address          50564 non-null  object
 3   Postcode         50564 non-null  object
 4   Easting          50564 non-null  int64 
 5   Northing         50564 non-null  int64 
 6   Latitude         50564 non-null  object
 7   Longitude        50564 non-null  object
 8   Local_authority  50564 non-null  object
dtypes: int64(3), object(6)
memory usage: 3.9+ MB


- after droping null values, total rows become 50564 and columns remains same.
- Latitude and Longitude dtype is object but it should be float.

In [12]:
df['Longitude']=df['Longitude'].astype(float)
df['Latitude']=df['Latitude'].astype(float)

C:\Users\HP\AppData\Local\Temp\ipykernel_2220\3280427824.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Longitude']=df['Longitude'].astype(float)
C:\Users\HP\AppData\Local\Temp\ipykernel_2220\3280427824.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Latitude']=df['Latitude'].astype(float)


In [14]:
df.to_csv(r"E:\Data Base\clean_open_pubs.csv",index=False)